libirods_rule_engine_plugin-unified_storage_tiering.cpp

```diff
@@ -84,12 +87,36 @@ namespace {
         return std::make_tuple(l1_idx, resource_name);
     } // get_index_and_resource
 
+    auto resource_hierarchy_has_good_replica(RcComm* _comm,
+                                             const std::string& _object_path,
+                                             const std::string& _root_resource) -> bool
+    {
+        namespace fs = irods::experimental::filesystem;
+
+        const auto object_path = fs::path{_object_path};
+
+        const auto query_string = fmt::format("select DATA_ID where DATA_NAME = '{0}' and COLL_NAME = '{1}' and "
+                                              "DATA_RESC_HIER like '{2};%' || = '{2}' and DATA_REPL_STATUS = '{3}'",
+                                              object_path.object_name().c_str(),
+                                              object_path.parent_path().c_str(),
+                                              _root_resource,
+                                              GOOD_REPLICA);
```

Comment:
Any reason why the value of `GOOD_REPLICA` isn't hard-coded into the query string?

Reply:
`GOOD_REPLICA` is '1' and so the query string would look like this if hard-coded:

`"select DATA_ID where DATA_NAME = '{0}' and COLL_NAME = '{1}' and DATA_RESC_HIER like '{2};%' || = '{2}' and DATA_REPL_STATUS = '1'"`

That confuses me at first glance because of the numbered parameters throughout the rest of the string, which makes me think it reads: `DATA_REPL_STATUS = '{1}'`. However, I could see it being confusing in the other direction, too: At a glance, it looks like we may be looking for objects with `DATA_REPL_STATUS = '3'`. :p

I'm happy to hard-code it if that is better.

Reply:
I like the abstraction.

Reply:
I don't have any strong preference for one over the other. I was just interested in the "why".

As a parameter, we may be paying the price of a few cycles depending on how optimizations go. However, in the grand scheme of things, the bottleneck will not be this code block.

_Sadly, it's clear I'm falling into the premature optimization hole again. Which is why I'm saying it's fine as is._

Reply:
I've hard-coded the value since the reason for using a parameter is equally weighted in either direction (i.e. `{3}` and `'3'` are just as easily visually confused (or not) as `'1'` and `{1}`).